src/rust/bitbox02-rust/src/hww/api/ethereum/address.rs

```diff
@@ -17,23 +17,39 @@ use sha3::digest::Digest;
 use alloc::string::String;
 use core::convert::TryInto;
 
+use super::pb;
+
 /// Generates a checksummed ethereum hex address from a 20 byte recipient.
 /// `recipient` - 20 byte tail (last 20 bytes of the pubkeyhash).
-pub fn from_pubkey_hash(recipient: &[u8; 20]) -> String {
+pub fn from_pubkey_hash(recipient: &[u8; 20], case: i32) -> String {
     let mut hex = [0u8; 40];
     hex::encode_to_slice(recipient, &mut hex).unwrap();
-    let hash = sha3::Keccak256::digest(&hex[..]);
-    for (i, e) in hex.iter_mut().enumerate() {
-        let hash_byte = {
-            let b = hash[i / 2];
-            if i % 2 == 0 {
-                b >> 4
-            } else {
-                b & 0xf
+    if case == pb::AddressCase::Upper as i32 { // upper case
+        for e in hex.iter_mut() {
+            if *e > b'F' {
+                *e -= 32; // convert to uppercase
+            }
+        }
+    } else if case == pb::AddressCase::Lower as i32 { // lower case
+        for e in hex.iter_mut() {
+            if *e > b'9' && *e < b'a' {
+                *e += 32; // convert to lowercase
+            }
+        }
+    } else { // take mixed case as default, case = pb::AddressCase::Mixed as i32
```

Comment:
I guess hex::encode() returns lowercase, so you don't need to do anything for to lowercase it?

Just use hex::encode_upper() instead if uppercase is needed for simplicity? Then we don't need be careful about if all the conditions here are correct. https://docs.rs/hex/latest/hex/fn.encode_upper.html


Also when you use `case: pb::EthAddressCase`, you can make a nice pattern match:

```
match case {
   EthAddressCase::EthAddressCaseLower => { ... }
    ... 
}
```